### ISSUE TITLE
[Feature Request] Add Anti-Leeching Functionality and Blocking Specific Clients or IPs to Address Malicious Download Behavior

### ISSUE THREAD

**Issue by zhiwenyyn** [transmission/transmission#6978]
### What is the issue?

Description：
Since the beginning of this year, a large number of leeching BT clients have suddenly appeared. They download without any upload behavior, leading to my upload bandwidth being consumed up to 500GB in a single day, eventually causing my broadband to be banned (in some regions, broadband upload traffic is limited). I hope Transmission can add anti-leeching functionality and the ability to block specific clients or IPs to counter this malicious downloading behavior.

For those who like to download and seed anime (and other content) locally using BT, please be aware that there is a client called github.com/anacrolix/torrent (and its renamed versions dt/torrent/v1.0*, hp/torrent/v2.0*), and similar clients like TrafficConsume and Taipei-torrent dev, that are massively requesting seeds and easily maxing out your upload bandwidth. Additionally, domestic ISPs are increasing their detection of PCDN(Peer to Peer Content Delivery Network), and large upload traffic may cause your broadband to be throttled or terminated. Therefore, it is recommended to take measures against this.

Current Information:
Who is doing this: Those involved in PCDN, and there are several groups. They have modified the original client to enable unlimited downloading and use methods like forging progress and changing client names to evade detection. Their purpose may be to balance PCDN upload and download traffic to evade ISP detection. Additionally, they seem to have updated and renamed their clients. The latest well-known client, representing the early anacrolix modification, is called xm/torrent/, with IP range 1.180.24.0/23.
Known malicious client names:

github.com/anacrolix/torrent
dt/torrent/v1.0 -DT0001-
hp/torrent/v2.0 -HP0001-
hp/torrent/v2.01
xm/torrent/ -XM0001-
Taipei-Torrent
taipei-torrent Dev
TrafficConsume
-GT0003-
There are also clients mimicking well-known ones, such as:

Transaction 2.93
Transaction 2.94
BitComet 2.04
Now some even mimic qBittorrent/4.6.5.

Affected range: Many torrents. These clients download almost any torrent, including but not limited to anime, movies, TV shows, software images, and various unknown torrents, with torrents larger than a few GB being more likely to be leeched (different groups target different torrents, such as TR 2.94 targeting software images).

Since when: Around early January 2024, and it is still ongoing and worsening.

Why they are doing this: From January 1, 2024, nationwide cross-province settlement of metropolitan network net outflow traffic began (in simple terms, upload ISPs pay, download ISPs earn), so ISPs started cracking down on PCDN and inspecting users with high upload traffic. One important indicator is the upload/download ratio; when the upload volume far exceeds the download volume, ISPs will throttle upload bandwidth, contact users, or even require offline inspections. Leechers then massively download to disguise the upload/download ratio, making it look normal to avoid ISP detection (reportedly only Unicom relies on this method). BT is the simplest, legal, and free way to max out download bandwidth, so it is targeted (some university mirror sites, JetBrains China, NPM, Mihoyo download links are also targeted).

Our uploads to these clients are useless. If Xunlei is leeching, these clients are bloodletting. Xunlei at least provides content and only leeches once, but these clients do not even write to the hard drive and max out bandwidth 24/7 (by discarding blocks, fake download progress, or using scripts to download repeatedly).

Victims: All seeders.

Why to prevent: ISPs are very strict on PCDN, cutting off high upload users. Large upload traffic may cause your broadband to be throttled or terminated. Downloaders also struggle to download torrents due to worse network conditions compared to leechers' data centers.

How to self-check and prevent: If you find a torrent constantly uploading, or specific episodes with high ratios of dozens to hundreds, or obscure torrents suddenly being heavily downloaded (especially if 3 or more users appear under v4/24 or v6/60 for the same torrent), or if the user list shows many identical clients, especially hp/torrent/v2.01 mentioned earlier, then you've been targeted.

Suggested Solutions:
Add anti-leeching functionality: Detect and limit or block these malicious leeching clients to protect users' upload bandwidth.
Blocking specific clients or IPs functionality: Allow users to manually add malicious clients or IPs to a blacklist to block their download requests.

### Which application of Transmission?

Web client

### Which version of Transmission?

Transmission 4.0.6

**Comment by Pentaphon** [comment]
I think a suitable solution that can reasonably be done is a "ban peer" function similar to qBittorrent. That is something that Transmission definitely needs in the long run to help protect users from leechers that exploit others. 

What you can do right now to avoid this is use the seed ratio limit and less obviously, stop seeding popular torrents.

**Comment by al3xtjames** [comment]
I am not familiar with this issue but [here's another discussion with some additional info](https://github.com/anacrolix/torrent/discussions/891).

**Comment by Cwek** [comment]
> I am not familiar with this issue but [here's another discussion with some additional info](https://github.com/anacrolix/torrent/discussions/891).

The discussion is relevant, but the point is that Transmission's manual peer management function is very poor. Qbittorrent can manually add peers or immediately terminate the connection with a peer and add its IP to the IP address blacklist to prevent it from connecting. Tixati also has similar capabilities. Transmission has an IP address blacklist, but after the list is updated, it cannot immediately terminate the connection of the corresponding IP address, and needs to restart the task to be effective.

The Developer of [PBH-BTN/PeerBanHelper](https://github.com/PBH-BTN/PeerBanHelper), which used to monitor the peer behavior of transmission and qbittorrent to prevent leeching peers, complained that the peer information of transmission is quite limited and difficult to control, and suggested using qbittorrent instead.

Transmission seems to be positioned as a lightweight and unmanaged bittorrent client, but "We are going to have to act if we want to live in a different world ."

**Comment by Avamander** [comment]
This is a big issue with Linux torrents (take any Ubuntu torrent for example). I'd love to see a built-in defense for this, but also support for external software (such as PeerBanHelper) to alleviate this problem.

**Comment by Avamander** [comment]
Yeah, this behaviour has become a significant drain on public torrents such as Ubuntu's .iso images. I had to stop seeding those torrents because there's still a significant waste of bandwidth before any external scripts are able to properly pick up on anomalous clients (without additional support from Transmission's side).
